components/builder-http-gateway/src/http/helpers.rs

```diff
@@ -496,12 +498,21 @@ fn do_group_promotion(
         package_ids.push(op.get_id());
     }
 
-    let mut opgp = OriginPackageGroupPromote::new();
-    opgp.set_channel_id(channel.get_id());
-    opgp.set_package_ids(package_ids);
-    opgp.set_origin(origin.to_string());
+    if promote {
+        let mut opgp = OriginPackageGroupPromote::new();
+        opgp.set_channel_id(channel.get_id());
+        opgp.set_package_ids(package_ids);
+        opgp.set_origin(origin.to_string());
 
-    route_message::<OriginPackageGroupPromote, NetOk>(req, &opgp)
+        route_message::<OriginPackageGroupPromote, NetOk>(req, &opgp)
+    } else {
+        let mut opgp = OriginPackageGroupDemote::new();
+        opgp.set_channel_id(channel.get_id());
+        opgp.set_package_ids(package_ids);
+        opgp.set_origin(origin.to_string());
+
+        route_message::<OriginPackageGroupDemote, NetOk>(req, &opgp)
+    }
```

Comment:
Could make this a little more general if `OriginPackageGroupPromote` and `OriginPackageGroupDemote` shared a trait covering these methods, but that seemed like overkill to me.

Reply:
There's a tradeoff to be struck between code that's overly generic and impossible to read, and not generic enough and repetitive/duplicated.  For a first pass at this, I think it's fine.